File: dotnet/Multiplayer/Rooms/Rooms.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Michitai.Multiplayer.Rooms
{
    /// <summary>
    /// Provides methods for game room management including creation, joining,
    /// player management, actions, updates, and administrative operations.
    /// </summary>
    public class Rooms
    {
        /// <summary>
        /// Creates a new game room with specified configuration.
        /// </summary>
        /// <typeparam name="TPlayerData">The type of player data for the host.</typeparam>
        /// <typeparam name="TRules">The type of game rules for the room.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's player token.</param>
        /// <param name="roomName">The name of the game room.</param>
        /// <param name="maxPlayers">Maximum number of players allowed (default: 4).</param>
        /// <param name="password">Optional password for the room.</param>
        /// <param name="hostSwitch">Whether host switching is allowed (default: false).</param>
        /// <param name="realtime">Whether the room supports realtime communication (default: false).</param>
        /// <param name="playerData">Optional player data for the host.</param>
        /// <param name="rules">Optional game rules for the room.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the created room ID.</returns>
        public static Task<RoomCreateResponse> CreateRoomAsync<TPlayerData, TRules>(Client client, string playerToken, string roomName, int maxPlayers = 4,
           string? password = null, bool hostSwitch = false, bool realtime = false, TPlayerData? playerData = null, TRules? rules = null, CancellationToken ct = default) where TPlayerData : class, new() where TRules : class, new()
           => client.Send<RoomCreateResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomCreate, $"&player_token={playerToken}"),
               new RoomCreateRequest<TPlayerData, TRules>(roomName, maxPlayers, password, hostSwitch, realtime, playerData, rules), ct);

        /// <summary>
        /// Retrieves a list of available game rooms.
        /// </summary>
        /// <typeparam name="T">The type to deserialize room rules into.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="search">Optional search term to filter rooms.</param>
        /// <param name="limit">Optional limit on the number of results.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the list of game rooms.</returns>
        public static Task<RoomListResponse<T>> GetRoomsAsync<T>(Client client, string? search = null, int? limit = null, CancellationToken ct = default) where T : class, new()
            => client.Send<RoomListResponse<T>>(HttpMethod.Post, client.Url(Endpoints.GameRoomList), new RoomListRequest { Search = search, Limit = limit }, ct);

        /// <summary>
        /// Joins an existing game room.
        /// </summary>
        /// <typeparam name="T">The type of player data.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="roomId">The ID of the room to join.</param>
        /// <param name="password">Optional password if the room is password-protected.</param>
        /// <param name="playerData">Optional player data to include when joining.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response confirming the player joined the room.</returns>
        public static Task<RoomJoinResponse> JoinRoomAsync<T>(Client client, string playerToken, string roomId, string? password = null, T? playerData = null, CancellationToken ct = default) where T : class, new()
            => client.Send<RoomJoinResponse>(HttpMethod.Post, client.Url(string.Format(Endpoints.GameRoomJoin, roomId), $"&player_token={playerToken}"),
                (password != null || playerData != null) ? new RoomJoinRequest<T>(password, playerData) : null, ct);

        /// <summary>
        /// Leaves the current game room.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response confirming the player left the room.</returns>
        public static Task<RoomLeaveResponse> LeaveRoomAsync(Client client, string playerToken, CancellationToken ct = default)
            => client.Send<RoomLeaveResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomLeave, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Gets the list of players in the current game room.
        /// </summary>
        /// <typeparam name="T">The type to deserialize player data into.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the list of players in the room.</returns>
        public static Task<RoomPlayersResponse<T>> GetRoomPlayersAsync<T>(Client client, string playerToken, CancellationToken ct = default) where T : class, new()
            => client.Send<RoomPlayersResponse<T>>(HttpMethod.Get, client.Url(Endpoints.GameRoomPlayers, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Sends a heartbeat to maintain the player's presence in the game room.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response confirming the heartbeat was received.</returns>
        public static Task<HeartbeatResponse> SendRoomHeartbeatAsync(Client client, string playerToken, CancellationToken ct = default)
            => client.Send<HeartbeatResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomHeartbeat, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Gets comprehensive information about the current game room including players and pending actions.
        /// </summary>
        /// <typeparam name="T">The type to deserialize room rules into.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing detailed room information.</returns>
        public  static Task<CurrentRoomResponse<T>> GetCurrentRoomAsync<T>(Client client, string playerToken, CancellationToken ct = default) where T : class, new()
            => client.Send<CurrentRoomResponse<T>>(HttpMethod.Get, client.Url(Endpoints.GameRoomCurrent, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Stops the current game room and removes all associated data (host only).
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's authentication token.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Success response confirming the room was stopped.</returns>
        public static Task<SuccessResponse> StopRoomAsync(Client client, string playerToken, CancellationToken ct = default)
            => client.Send<SuccessResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomStop, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Kicks a player from the game room (host only).
        /// Cannot kick yourself.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's authentication token.</param>
        /// <param name="playerId">The ID of the player to kick.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response confirming the player was kicked.</returns>
        public static Task<RoomKickResponse> KickPlayerAsync(Client client, string playerToken, int playerId, CancellationToken ct = default)
            => client.Send<RoomKickResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomKick, $"&player_token={playerToken}"), new RoomKickRequest { Player_id = playerId }, ct);

        /// <summary>
        /// Updates the password for the game room (host only).
        /// Use empty string to remove the password.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The host's authentication token.</param>
        /// <param name="password">New password, or null to remove the password.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Success response confirming the password was updated.</returns>
        public static Task<SuccessResponse> UpdateRoomPasswordAsync(Client client, string playerToken, string? password = null, CancellationToken ct = default)
            => client.Send<SuccessResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomPassword, $"&player_token={playerToken}"), new RoomPasswordUpdateRequest(password), ct);
    }
}

File: python/multiplayer/rooms/rooms.py
"""Game room management: create, join, leave, players, heartbeat, admin ops."""

from __future__ import annotations

import dataclasses
from typing import Any, Optional, Type, TypeVar

from .._serialization import from_dict
from ..client import Client
from .models import (
    CurrentRoomResponse,
    HeartbeatResponse,
    RoomCreateResponse,
    RoomJoinResponse,
    RoomKickResponse,
    RoomLeaveResponse,
    RoomListResponse,
    RoomPasswordResponse,
    RoomPlayersResponse,
    RoomStopResponse,
)

T = TypeVar("T")

# Endpoints
GAME_ROOM_CREATE = "game_room.php/create"
GAME_ROOM_LIST = "game_room.php/list"
GAME_ROOM_JOIN = "game_room.php/{room_id}/join"
GAME_ROOM_LEAVE = "game_room.php/leave"
GAME_ROOM_PLAYERS = "game_room.php/players"
GAME_ROOM_HEARTBEAT = "game_room.php/heartbeat"
GAME_ROOM_CURRENT = "game_room.php/current"
GAME_ROOM_STOP = "game_room.php/stop"
GAME_ROOM_KICK = "game_room.php/kick"
GAME_ROOM_PASSWORD = "game_room.php/password"


class Rooms:
    """Game room lifecycle and player management."""

    def __init__(self, client: Client) -> None:
        self._client = client

    def create(
        self,
        player_token: str,
        room_name: str,
        max_players: int = 4,
        password: Optional[str] = None,
        host_switch: bool = False,
        realtime: bool = False,
        player_data: Any = None,
        rules: Any = None,
    ) -> RoomCreateResponse:
        """Create a new game room.

        :param player_token: The host's player token.
        :param room_name: Name of the game room.
        :param max_players: Maximum players allowed (default 4).
        :param password: Optional room password.
        :param host_switch: Whether host switching is allowed.
        :param realtime: Whether the room supports realtime communication.
        :param player_data: Optional player data for the host.
        :param rules: Optional game rules for the room.
        """
        return self._client.send(
            "POST",
            self._client.player_url(GAME_ROOM_CREATE, player_token),
            {
                "room_name": room_name,
                "max_players": max_players,
                "password": password,
                "host_switch": host_switch,
                "realtime": realtime,
                "player_data": player_data,
                "rules": rules,
            },
            RoomCreateResponse,
        )

    def list(
        self,
        search: Optional[str] = None,
        limit: Optional[int] = None,
        rules_cls: Optional[Type[T]] = None,
    ) -> RoomListResponse:
        """List available game rooms.

        :param rules_cls: Optional dataclass type for each room's ``rules``.
        """
        response = self._client.send(
            "POST",
            self._client.url(GAME_ROOM_LIST),
            {"search": search, "limit": limit},
            RoomListResponse,
        )
        _apply_type((r for r in response.rooms), "rules", rules_cls)
        return response

    def join(
        self,
        player_token: str,
        room_id: str,
        password: Optional[str] = None,
        player_data: Any = None,
    ) -> RoomJoinResponse:
        """Join an existing game room."""
        body = None
        if password is not None or player_data is not None:
            body = {"password": password, "player_data": player_data}
        return self._client.send(
            "POST",
            self._client.player_url(
                GAME_ROOM_JOIN.format(room_id=room_id), player_token
            ),
            body,
            RoomJoinResponse,
        )

    def leave(self, player_token: str) -> RoomLeaveResponse:
        """Leave the current game room."""
        return self._client.send(
            "POST",
            self._client.player_url(GAME_ROOM_LEAVE, player_token),
            None,
            RoomLeaveResponse,
        )

    def get_players(
        self,
        player_token: str,
        data_cls: Optional[Type[T]] = None,
    ) -> RoomPlayersResponse:
        """Get the list of players in the current room.

        :param data_cls: Optional dataclass type for each player's ``player_data``.
        """
        response = self._client.send(
            "GET",
            self._client.player_url(GAME_ROOM_PLAYERS, player_token),
            None,
            RoomPlayersResponse,
        )
        _apply_type((p for p in response.players), "player_data", data_cls)
        return response

    def heartbeat(self, player_token: str) -> HeartbeatResponse:
        """Send a heartbeat to maintain presence in the room."""
        return self._client.send(
            "POST",
            self._client.player_url(GAME_ROOM_HEARTBEAT, player_token),
            None,
            HeartbeatResponse,
        )

    def get_current(
        self,
        player_token: str,
        rules_cls: Optional[Type[T]] = None,
    ) -> CurrentRoomResponse:
        """Get detailed info about the current room incl. pending items.

        :param rules_cls: Optional dataclass type for the room's ``rules``.
        """
        response = self._client.send(
            "GET",
            self._client.player_url(GAME_ROOM_CURRENT, player_token),
            None,
            CurrentRoomResponse,
        )
        if rules_cls is not None and response.room is not None:
            response.room.rules = _convert(response.room.rules, rules_cls)
        return response

    def stop(self, player_token: str) -> RoomStopResponse:
        """Stop the current game room and remove all its data (host only)."""
        return self._client.send(
            "POST",
            self._client.player_url(GAME_ROOM_STOP, player_token),
            None,
            RoomStopResponse,
        )

    def kick(self, player_token: str, player_id: int) -> RoomKickResponse:
        """Kick a player from the room (host only). Cannot kick yourself."""
        return self._client.send(
            "POST",
            self._client.player_url(GAME_ROOM_KICK, player_token),
            {"player_id": player_id},
            RoomKickResponse,
        )

    def update_password(
        self,
        player_token: str,
        password: Optional[str] = None,
    ) -> RoomPasswordResponse:
        """Update the room password (host only). Pass ``None`` to remove it."""
        return self._client.send(
            "POST",
            self._client.player_url(GAME_ROOM_PASSWORD, player_token),
            {"password": password},
            RoomPasswordResponse,
        )


def _convert(value: Any, cls: Type[T]) -> Any:
    if dataclasses.is_dataclass(cls) and isinstance(value, dict):
        return from_dict(cls, value)
    return value


def _apply_type(items: Any, attr: str, cls: Optional[Type[T]]) -> None:
    if cls is None:
        return
    for item in items:
        setattr(item, attr, _convert(getattr(item, attr), cls))
